Fix #775: update HTMLAllCollection to match reality better

domenic opened the pull request:
As discussed extensively in that issue, in most browsers
HTMLAllCollection is more lenient than previously believed, especially
with regard to its item() method and legacycaller behavior.

@DigiTec and @nox to review, plus @annevk if he wants.

DigiTec commented:
Adding @travisleithead as well since we had an extensive discussion this morning. Consider me pretty much signed off.

nox commented on source:
```diff
@@ -7019,9 +7017,9 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
-  (<span>HTMLCollection</span> or <span>Element</span>)? <span data-x="dom-HTMLAllCollection-item-string">item</span>(DOMString name);
-  legacycaller getter (<span>HTMLCollection</span> or <span>Element</span>)? <span data-x="dom-HTMLAllCollection-namedItem">namedItem</span>(DOMString name);
+  getter <span>Element</span>? (unsigned long index);
+  getter (<span>HTMLCollection</span> or <span>Element</span>)? namedItem(DOMString name);
```
Why did you remove the `<span>` around `namedItem`?

domenic replied:
Fixed, thanks!

nox commented:
Just one question, otherwise SGTM.

annevk commented:
@smaug---- could you take a look at this?

annevk commented on source:
```diff
@@ -7032,13 +7030,14 @@ interface <dfn>HTMLAllCollection</dfn> {
 
-   <dt><var>element</var> = <var>collection</var> . <code subdfn data-x="dom-HTMLAllCollection-item-string">item</code>(<var>name</var>)</dt>
-   <dt><var>collection</var> = <var>collection</var> . <code data-x="dom-HTMLAllCollection-item-string">item</code>(<var>name</var>)</dt>
+   <dt><var>element</var> = <var>collection</var> . <code data-x="dom-HTMLAllCollection-item">item</code>(<var>name</var>)</dt>
```
Why remove the subdfn annotation? Isn't that important for dev edition? (Not sure, but that's my hunch.)

domenic replied:
Because the build script barfs if you have more than one subdfn with the same data-x. (line 7032 has the other one)

zcorpan replied:
`document.all` itself is in the obsolete section. Maybe this whole section should not be in the dev edition?

domenic replied:
Nah. Other obsolete APIs are in dev. If it's in MDN it should be in dev, IMO. Besides, infoboxes are useful (although sometimes problematic) for non-devs.

foolip commented on source:
```diff
@@ -7062,6 +7061,44 @@ interface <dfn>HTMLAllCollection</dfn> {
+
+  <ol>
+
+   <li><p>If <var>nameOrIndex</var> was not provided, return null.</p></li>
```
When matching Blink's IDL to what's proposed here, this bit became very artificial-looking, a separate function for the no-argument case which only returns null. We could add a use counter to see if it's safe to make the argument non-optional, or do you think we should just go with the maximum non-throwing union and not worry about it?

nox replied:
> If multiple legacy callers are specified on an interface, overload resolution is used to determine which legacy caller is invoked when the object is called as if it were a function.

Seems feasible.

domenic replied:
I don't understand the concern. Is the idea that Blink doesn't support optional correctly and so you want to turn this into an overload?

In general I think if we were to go down the route of use-countering this into a minimal subset, we'd end up removing a lot more than just the no-arg case.

foolip replied:
I don't think there's any reason to strive for a minimal subset, it's just that the optionality of this argument looks mysterious when the spec's IDL is used.

foolip replied:
And since `document.all()` and `document.all.item()` never return anything useful, it's very likely that the argument can be safely made non-optional.

domenic replied:
I still don't understand the concern. It doesn't look mysterious to me (anymore than this entire API is a mysterious clusterfuck of legacy garbage). It's optional. What's mysterious about that?

zcorpan commented on source:
```diff
@@ -7062,6 +7061,54 @@ interface <dfn>HTMLAllCollection</dfn> {
+
+    <ol>
+     <li><p>If <var>nameOrIndex</var> is a <span>valid non-negative integer</span> that represents
+     a number less than 2<sup>32</sup> - 1, return the result of <span
```
I suppose this is simple enough to be interoperably implemented, but I think in general the spec explicitly invokes the parse algorithms instead of handwaving "represents". Is there a reason to not spec this more like how you would implement it?

I found while writing the below that the removed text uses "greater than 2<sup>32</sup> - 1", but the new text uses "less than 2<sup>32</sup> - 1"; shouldn't it be "less than 2<sup>32</sup>" instead?

---

1. If *nameOrIndex* is a valid non-negative integer, follow these substeps:
    1. Let *index* be the result of applying the rules for parsing non-negative integers to *nameOrIndex*.
    1. If *index* is less than 2<sup>32</sup>, return the result of getting the "all"-indexed element from this `HTMLAllCollection` given *index*, and abort these steps.
1. Return the result of getting the "all"-named element(s) from this `HTMLAllCollectio` given *nameOrIndex*.

domenic replied:
represents is used elsewhere for this purpose, e.g. "A valid non-negative integer that is exactly two digits long and that represents a number in the range 00..23.". It seems simpler and less repetitive. I imagine implementations will implement this more like my current wording, actually: Chrome at least does basically `int* result; if (toInt(&string)) { ... use *int ... }`. Here `*int` is "the number represented by nameOrindex".

Good catch on less than.

zcorpan replied:
Very well :-)

domenic commented:
@DigiTec in the latest commit I changed it to just use a DOMString for item/legacycaller, and then do number parsing, since the unsigned long codepath was redundant. If you could give it a look-over that'd be helpful.

At this point I think I'd like LGTMs from @DigiTec and @foolip before merging. Discussion has settled a bit.

foolip commented on source:
```diff
@@ -7062,6 +7061,43 @@ interface <dfn>HTMLAllCollection</dfn> {
   <p>The object's <span>supported property indices</span> are as defined for
   <code>HTMLCollection</code> objects.</p>
 
+  <p>The <span>supported property names</span> consist of the non-empty values of all the <code
```
Double-checked that this paragraph didn't change at all.

foolip commented on source:
```diff
@@ -7062,6 +7061,43 @@ interface <dfn>HTMLAllCollection</dfn> {
+  <ol>
+   <li><p>If <var>nameOrIndex</var> was not provided, return null.</p></li>
+
+   <li><p>If <var>nameOrIndex</var> is a <span>valid non-negative integer</span> that represents
```
This doesn't allow a leading `+`, but all of the conversion methods I found in Blink would allow it, so an extra check was needed. A note in the spec is probably overly pedantic, but this is something to remember when writing or reviewing tests for this.

nox replied:
I don't think there is a need for a note at all since:

> A string is a **valid non-negative integer** if it consists of one or more ASCII digits.

https://html.spec.whatwg.org/multipage/#valid-non-negative-integer

nox commented on source:
```diff
@@ -7062,6 +7061,43 @@ interface <dfn>HTMLAllCollection</dfn> {
+   <li><p>If <var>nameOrIndex</var> was not provided, return null.</p></li>
+
+   <li><p>If <var>nameOrIndex</var> is a <span>valid non-negative integer</span> that represents
+   a number less than 2<sup>32</sup>, return the result of <span
```
https://html.spec.whatwg.org/multipage/#limited-to-only-non-negative-numbers-greater-than-zero

Shouldn't we be consistent and use the number 4294967296 here instead of a mathematical expression?

domenic replied:
Ah, yeah I guess we should, thanks. Will update.

foolip commented on source:
```diff
@@ -7082,53 +7118,37 @@ interface <dfn>HTMLAllCollection</dfn> {
-    <p>Let <var>collection</var> be an <code>HTMLCollection</code> object rooted at the
-    same <code>Document</code> as the <code>HTMLAllCollection</code> object on which the method was
-    invoked, whose filter matches only elements that are either:</p>
+    <p>Let <var>subCollection</var> be an <code>HTMLCollection</code> object rooted at the same
```
I wouldn't have guessed it, but this collection really is live in Edge:
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3949

That's a separate discussion, though.

bzbarsky replied:
It's live in Firefox too, and per spec, yes?

foolip replied:
Yep, it's live in Firefox too, as per spec:
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3968

foolip commented:
This now LGTM, certainly a huge improvement compared to before.

DigiTec commented:
HTMLCollection is always live and the spec has been that way for a long time as has Edge and before that all the way back to IE 5. Note that many other collections, like getElementsByTagName, etc... also have this. As do all window group collections like links and anchors. This was the NodeList vs HTMLCollection discussion very early in the original thread.

DigiTec commented:
I only have one remaining point of contention.

console.log(document.all.item(-2));

Edge doesn't convert this to "-2". The reason is that we obey the unsigned long conversion first since -2 is a number. I realize overnight as you guys were figuring out the string/number parsing semantics across the browsers that this led, due to Chrome's behavior, of a way to use DOMString and then parse a positive number.

At this point both Edge/FireFox do agree here that -2 is a number and its an invalid index and so it returns null (or errantly undefined in the case of legacycaller).

I actually liked all of the discussion and rules for (unsigned long or DOMString) and I think they were great. I really think the right solution is to use the union type and then have special rules when processing a resulting DOMString to note that it could be a numeric string which requires additional parsing. What use case does having this break? Right now the current webIDL breaks tons of FireFox/Edge behavior.

domenic commented:
> I really think the right solution is to use the union type and then have special rules when processing a resulting DOMString to note that it could be a numeric string which requires additional parsing. What use case does having this break?

It breaks no use case, but it's completely pointless, because anything you can spec this way, you can spec with fewer steps by just accepting a string instead of a union type and then processing the strings accordingly.

The only way a union type would make sense is if browsers treated one of the numbers in the range [0, 2^32] differently from the corresponding quoted string. That would mean using something besides `unsigned long` in the union. But not even Edge or Firefox seems to do that.

> Right now the current webIDL breaks tons of FireFox/Edge behavior.

So, I think the current IDL is the best cross-section for all browsers. But you seem to be proposing switching from `unsigned long` to something that allows negative numbers, maybe? I don't think this is a majority position, but do you feel strongly it should be handled like that?

DigiTec commented:
Just because it says unsigned long, doesn't mean that it fails to parse -2 as an unsigned long and then falls back to the DOMString version. That is my point. unsigned long means, ANY numeric type. If it is out of range, then it is out of range, And a negative collection index should return null, not be treated like "-2" and then return the element of that name instead.

That is why it is different. You don't need a more permissive numeric type, you just need to evaluate all numbers as unsigned long and if they are out of range treat them in their out of range condition.

domenic commented:
> Just because it says unsigned long, doesn't mean that it fails to parse -2 as an unsigned long and then falls back to the DOMString version. That is my point. unsigned long means, ANY numeric type.

No, unsigned long has well specified behavior. If you pass -2 to it, it will (before it ever reaches the HTML spec) be converted to 2^32 - 2. There is no "out of range" path where you get to choose the behavior. You need to use a type that allows the range you want.

domenic commented:
BTW if you'd like to discuss this in real-time feel free to hop in IRC: https://wiki.whatwg.org/wiki/IRC

domenic commented:
If you want "any numeric type" then you probably want `unrestricted double` instead.

DigiTec commented:
https://heycam.github.io/webidl/#es-unsigned-long

domenic commented:
Yep, see step 4.

domenic commented:
compare to https://heycam.github.io/webidl/#es-unrestricted-double

DigiTec commented:
My point is that -2 is still a number. We actually do right now exactly what the spec says. So compare our behavior. 

Because of unsigned long, we MUST CONVERT -2 into a rather odd value and then have it fail to index. What I"m saying is that is different from the DOMString conversion to positive value that is being proposed.

Why would we not obey the ToUInt32 conversion semantics and still treat it like an index and not like a string for the -2 case?

domenic commented:
I see, so you're saying that it should be treated as 2^32 - 2 instead of as the ID "-2"?

I think that's a possible position to take on this issue, but it's another 2/2 split with the mobile browser majority preferring the ID "-2". I don't particularly care, and you're the vendor who's most engaged (which is more important than marketshare for such likely-not-a-web-compat-issue things). So if you'd prefer to go with the Edge/Firefox side of the split instead of the Chrome/Safari side, then we can spec that. Just confirm to me that's what you want and we can revise the PR.

DigiTec commented:
What I can offer is that we've yet to see any compat issues from our more restrictive behavior. That doesn't mean we don't have something sitting in a reduction queue, but I haven't seen it. Since we are already doing the more restrictive behavior, I can make the upgrades on our side and keep that in place (achieve full convergence with the spec without relaxing too much). If we combine this with your use counters and you are able to show compat hit with them, then we could then open it up to the less restrictive approach.

I think just like @foolip doesn't necessarily want to see all() and item() get opened up (the optional behavior) I'm equivalently worried about opening up the string-to-number approach.

I'm more than willing to relax our behavior in the future if your use counters support that decision. And if we hit any compat issues we'll report those out immediately.

domenic commented:
Heh, I guess that's where we differ. I've completely given up on making this API rational and have instead settled on trying to spec something that is maximally compatible. Whereas you have rationality concerns about negative numbers (and even about legacycaller vs. item), and @foolip has rationality concerns about the undefined/no-args. I really don't think that's worthwhile since this API is such a legacy clusterfuck anyway. But I'm happy to make you happy so I'll make the change.

I'll ping this thread when that's done.

domenic commented:
OK, let's try reviewing again @DigiTec @foolip @nox? I've reverted back to the previous version and added a note showing the consequences for numeric cases, which I'll paste here:

```js
document.all(1)            // looks up the element at index 1
document.all("1")          // looks up the element at index 1

document.all(-2)           // looks up the element at index 4294967294 (= 4294967296 - 2)
document.all("-2")         // looks up the element with ID or name "-2"

document.all(0x3)          // looks up the element at index 3
document.all("0x3")        // looks up the element with ID or name "0x3"

document.all(4294967296)   // looks up the element at index 0
document.all("4294967296") // looks up the element with ID or name "4294967296"
```

DigiTec commented:
Okay, so per your test cases legacycaller in the case of "1" will still do the internal "1" to 1 conversion. That is in prose, I assume. Since .item("1") doesn't have the special parsing behavior? Sorry, I'm in meetings for the rest of the day, so my responses and questions will lag.

```javascript
console.log(document.all("1")); // Returns head, looks up item at index 1 due to string to int conversion
console.log(document.all.item("1")); // Returns null due to no element with id "1"
```

domenic commented:
Yep, that's in prose.

foolip commented:
So much happens while you sleep! I did some testing in https://github.com/whatwg/html/issues/775#issuecomment-192154336 and https://github.com/whatwg/html/issues/775#issuecomment-192171566.

I've tried implementing the latest proposal in Chromium (with `unsigned long or DOMString` + in-prose parsing) and the results seem kind of weird. Here's the output from my tests:
```
log: document.all["0"] => <html>
log: document.all("0") => <html>
log: document.all.item("0") => <html>
log: 
log: document.all["00"] => <div id="00">
log: document.all("00") => <html>
log: document.all.item("00") => <html>
log: 
log: document.all["1"] => <head>
log: document.all("1") => <head>
log: document.all.item("1") => <head>
log: 
log: document.all["01"] => <div id="01">
log: document.all("01") => <head>
log: document.all.item("01") => <head>
log: 
log: document.all["-1"] => <div id="-1">
log: document.all("-1") => <div id="-1">
log: document.all.item("-1") => <div id="-1">
log: 
log: document.all["+1"] => <div id="+1">
log: document.all("+1") => <div id="+1">
log: document.all.item("+1") => <div id="+1">
log: 
log: document.all["100"] => undefined
log: document.all("100") => null
log: document.all.item("100") => null
log: 
log: document.all["2147483647"] => undefined
log: document.all("2147483647") => null
log: document.all.item("2147483647") => null
log: 
log: document.all["2147483648"] => undefined
log: document.all("2147483648") => null
log: document.all.item("2147483648") => null
log: 
log: document.all["4294967294"] => undefined
log: document.all("4294967294") => null
log: document.all.item("4294967294") => null
log: 
log: document.all["4294967295"] => <div id="4294967295">
log: document.all("4294967295") => null
log: document.all.item("4294967295") => null
log: 
log: document.all["4294967296"] => <div id="4294967296">
log: document.all("4294967296") => <div id="4294967296">
log: document.all.item("4294967296") => <div id="4294967296">
```
and
```
log: document.all[0] => <html>
log: document.all(0) => <html>
log: document.all.item(0) => <html>
log: 
log: document.all[1] => <head>
log: document.all(1) => <head>
log: document.all.item(1) => <head>
log: 
log: document.all[-1] => <div id="-1">
log: document.all(-1) => null
log: document.all.item(-1) => null
log: 
log: document.all[-2] => <div id="-2">
log: document.all(-2) => null
log: document.all.item(-2) => null
```
The concern in https://github.com/whatwg/html/issues/775#issuecomment-191917461 was with `document.all.item(-1)`, but it seems to me that `document.all[-1]` is just as likely a construct.

This brings us to the fact that `document.all[-1]` finds the `<div id="-1">` element in all engines except Edge. The rules for when to do `ToUint32` and indexed property lookup must differ, but what's the right behavior?

It would be *nice* if `document.all[x] === document.all(x) === document.all.item(x)` for all *x*, but I don't think that's achievable, so it's just a matter of picking where and how things are inconsistent. This is all pretty broken already, so almost any permutation of behavior should work :)

Not sure what to suggest, care to investigate further, @DigiTec?

bzbarsky commented:
My take on this is that absent known compat issues, we should spec the simplest thing we can here, which is what the current spec was aiming at.  I don't think we should be doing the weird stringify-and-reparse thing at all unless there are concrete cases where it's actually needed, especially since it requires custom, error-prone, code, and slows down indexed access.

> This brings us to the fact that document.all[-1] finds the <div id="-1"> element in all engines except Edge. 

It doesn't in Firefox.

bzbarsky commented:
> It doesn't in Firefox.

Er, I was wrong.  It does; for some reason this doesn't work for `<html id="-1">`...

DigiTec commented:
@foolip Our behavior for getters is based on how the JS engine calls us. We have two methods that they call us on. GetOwnProperty and GetOwnItem. The later is generally used when we see the bracket indexing [] and we get the underlying JS object that was passed in as the index (so we get a variable type and we can coerce it as necessary). We then call ToInt on that value without any additional coercion today.

The other variation where we have GetProperty which means a string was passed in we actually get an atomized property id. This we get the name from. Then we try to convert it to a number. If the conversion succeeds we use the index, if not we use the string. This is very high level I need to dig into the specifics.

I think our goal was to be as close to actual JavaScript semantics as possible (initially) and then where we needed to adapt that a bit to be more HTML like in our indexing behavior. JavaScript treats numeric and string indexes that look like numerics, as numerics AFAICT. Simple test case below. For non array objects all indices get converted to property names instead.

The issue with our collections is that due to the platform object dynamic namespace backing them (the DOM can change) and the fact that id's and names can conflict with ordinal indices, we aren't purely an array nor purely an object and thus our getters at some level have to disambiguate.

```javascript
var arr = [];

arr[-2] = "foo";
arr[1] = "bar";
arr["1"] = "baz";

console.log(JSON.stringify(arr), arr[-2], arr["-2"], arr[1], arr["1"]);

var obj = {};
obj[-2] = "foo";
obj[1] = "bar";
obj["1"] = "baz";

console.log(JSON.stringify(obj), obj[-2], obj["-2"], obj[1], obj["1"]);
```

**Chrome, Edge, FireFox**
```
[null,"baz"] foo foo baz baz
(index):58 {"1":"baz","-2":"foo"} foo foo baz baz
```

https://jsfiddle.net/z7d1srwL/

DigiTec commented:
**Edge Verified**: If you use getter syntax and the collection is array-like (meaning it can accept ordinals) then we are most likely (I haven't verified all overloads) going to convert ToInt and then verify range.

This creates a state where `all[]` and `item()` definitely return different results. 

It also creates a situation where `all[]` and `all()` can return different results though I haven't crafted it yet. Basically if JavaScript can convert a variable to a number that is usable but we can't convert the string representation then the two could diverge. Given our range restrictions this doesn't seem to be anything to worry about.

I guess this type is really messed up across all the browsers. How we treat legacycaller vs getter vs item, we all seem to have slightly different mappings. We also all seem to have slightly different conversions on the index/parameter.

foolip commented:
@bzbarsky 
> absent known compat issues, we should spec the simplest thing we can

The other goal is interoperability. In a case like `document.all(0)` where 3/4 engines agree on a sane behavior, it seems like the obvious thing to do.

For `document.all("0")` there are a few lines of code that could be avoided by nuking it, but it's still mostly harmless.

If there are specific changes you don't want, or all of them, I can try to probe for them with use counters. It would be nice to be able to move faster on issues like this by tending towards the union of all semi-sane behaviors, so if there is any implementor who'd like to go ahead without waiting for telemetry, I think we should value that momentum over simplicity of the spec.

bzbarsky commented:
> In a case like document.all(0) where 3/4 engines agree on a sane behavior, it seems like the obvious thing to do.

Yes, that seems fine.

> For document.all("0") there are a few lines of code that could be avoided by nuking it

I disagree.  Now you're talking string-to-integer parsing, which is an interop can of worms unless you're really careful to define exactly how it's going to work and what all the edge cases are and how they get treated.  I mean, this is a situation in which even the ES6 spec is not consistent with itself (for historical reasons, but still)...

Note that of the three independent implementations here (I'm counting WebKit and Blink as non-independent, because this bit predates the fork), two do not support `document.all("0")`...

foolip commented:
> Note that of the three independent implementations here (I'm counting WebKit and Blink as non-independent, because this bit predates the fork), two do not support `document.all("0")`...

Edge supports it too, in my testing with http://software.hixie.ch/utilities/js/live-dom-viewer/saved/3956 at least. It doesn't support `document.all.item("0")` in the same way, however, that's Blink/WebKit only.

DigiTec commented:
This is what I can now verify. For all Edge **getters** we are relying on ES 6 behavior for determining if something is an index or a property. However, we are messing up in the "property" case and doing an additional numeric conversion (unnecessary) which causes us to treat things like "-2" as -2 rather than keeping them as strings.

Oddly, I think this is legacy for us where we didn't evolve with Chakra and we now have redundant code in place.

Our legacycaller goes through the "property" case which does string -> integer conversion. I think the right behavior for Edge would be to:

1. When Chakra tells us its an index, we only do indexing.
2. When Chakra tells us its a property, we only do identifier lookup, no more fallback to numeric conversion.
3. legacycaller ends up being whatever we define here in the spec.
4. item ends up being whatever we define here in the spec.

I would favor the `(unsigned long or DOMString)` using the WebIDL semantics we already discussed for `item`. 

If we agreed to that then the only remaining contentious point would be if we want legacycaller to have extra semantics. 

I believe `item` would play like this:
1. If nameOrIndex is a numeric type then ToInt32 an index
2. Otherwise use ToString and perform identifier lookup

We could say that legacycaller gets the extra bits which is this:
1. If nameOrIndex is a numeric type then ToInt32 and index
2. Otherwise ToString and try to parse as an integer
2.1 If the resulting number is a valid positive integer index
2.2 Else treat like a string and do identifier indexing.

However, we could also simplify and make legacycaller and item behave exactly the same.

DigiTec commented:
I wanted to split 1 issue out separately at this point. 

```javascript
all("0")
item("0")
```

During the discussion it has come up that [], () and item() could potentially behave the same. My question is this. Given that the binding layer for getters in ES is what gets us the correct indexer behavior and we want the others to behave the same, is it not true that we can't do that? After all there is no type we can define or behavior we can define other than having an algorithm run in the method itself that would do the "is this a valid ES index" in this context.

domenic commented:
> If nameOrIndex is a numeric type then ToInt32 an index

Can you be a bit more precise? If it's `unsigned long or DOMString`, then a ToUint32 is already done in the unsigned long conversion, and ToInt32 will do nothing. Are you maybe proposing changing the type to `long`? In that case there's still no need for ToInt32 since that's taken care of by the type signature, so any time it's a numeric type it'll already be ToInt32ed.

> However, we could also simplify and make legacycaller and item behave exactly the same.

That is what the current spec in this pull request does. It seems nicer to me...

> During the discussion it has come up that [], () and item() could potentially behave the same. My question is this. Given that the binding layer for getters in ES is what gets us the correct indexer behavior and we want the others to behave the same, is it not true that we can't do that?

I think this could be done by changing legacycaller and item() to take `any` and have them do in prose `Get(this, arg)`. This would then delegate to the anonymous indexed string and unsigned long getters.

However, this will change things observably (e.g. it will trigger getters defined in the prototype chain) and probably would not make @bzbarsky happy since (I believe) it requires custom bindings.

bzbarsky commented:
> and have them do in prose `Get(this, arg)`

That's sort of what I was trying to get at in https://github.com/whatwg/html/issues/775#issuecomment-192403572

You're right that it's observably different from doing manual parsing, of couse.  Not just because of getters on the proto chain but also if someone does `document.all(Symbol.iterator)` and the like (yes, I know....).

> since (I believe) it requires custom bindings

Fwiw, in Gecko it doesn't require those.  Functions that take `any` get enough stuff passed in through the binding layer that they can operate sanely on the `any`, which includes enough stuff to do a `Get(this, arg)` in this case.

DigiTec commented:
> Can you be a bit more precise? If it's unsigned long or DOMString, then a ToUint32 is already done in 

Sorry, I simply misspoke, Replace ToInt32 with ToUInt32. I'm just basically saying that, if it is numeric as detected by type, then treat it as the WebIDL type which in this case is unsigned long and do the conversion that we discussed before. That -2 is actually the whole modulus thing written in the spec, bu it remains a number between 0 and UINT32_MAX. I think we agreed previously that this is how the union type would work, if we read WebIDL conversion rules, with no additional prose applied to support more obscure behaviors.

> However, this will change things observably (e.g. it will trigger getters defined in the prototype chain) and probably would not make @bzbarsky happy since (I believe) it requires custom bindings.

I agree with that too. I was throwing it out there as a, "what are we kind of trying to do" and "does it make any sense". I think in binding terms the answer is, "it doesn't make sense, so let's stick to bindings".

> That is what the current spec in this pull request does

Okay, the last version I read, I thought, still had special rules in the prose that allowed a string to be converted back to an index in the legacycaller case. I'll go back and review again ;-)

DigiTec commented:
Another thing to point out here I think is that once we lock this behavior down, there are parts of our discussion that go into HTMLCollection and NodeList. Since we'd adopt the ES behavior of allowing the bindings to decide if an index is numeric or string and stop doing our special parsing. I don't think it changes any prose or WebIDL in those cases, but I bet it does have implications on pass rates for test cases targeting the fringe behaviors.

domenic commented:
> Sorry, I simply misspoke, Replace ToInt32 with ToUInt32

Ah OK cool, all good then.

> Okay, the last version I read, I thought, still had special rules in the prose that allowed a string to be converted back to an index in the legacycaller case. I'll go back and review again ;-)

Oh, it does. It's just the same rules are used for legacycaller as for index. They both go down your "We could say that legacycaller gets the extra bits which is this:" path.

> Another thing to point out here I think is that once we lock this behavior down, there are parts of our discussion that go into HTMLCollection and NodeList. Since we'd adopt the ES behavior of allowing the bindings to decide if an index is numeric or string and stop doing our special parsing. I don't think it changes any prose or WebIDL in those cases, but I bet it does have implications on pass rates for test cases targeting the fringe behaviors.

Hmm, I don't quite understand what you mean here. HTMLCollection and NodeList are pretty well-defined, mainly because they don't have any API surface that accepts a nameOrIndex; it's always unambiguous and the bindings should take care of routing any incoming values to the right place.

DigiTec commented on source:
```diff
@@ -7062,6 +7061,76 @@ interface <dfn>HTMLAllCollection</dfn> {
+     data-x="concept-get-all-indexed">getting the "all"-indexed element</span> from this
+     <code>HTMLAllCollection</code> given the number represented by <var>nameOrIndex</var>.</p></li>
+
+     <li><p>Otherwise, return the result of <span data-x="concept-get-all-named">getting the
```
In our comment thread, I was kind of making the point of whether or not we could remove this fallback altogether. It seems like we believe the right approach is either:
1. Have the fallback and use it in both legacycaller and item (unified), matches Chrome today
2. Not have the fallback at all (still unified), matches FireFox today

Edge only supports the string prose on all("1") today so I'm open to aligning with FireFox rather than expanding the item("1") case? Maybe? Thoughts? If so, this entire prose goes away.

Per @foolip tests:
**FireFox**
```
log: document.all["1"] => <head>
log: document.all("1") => <div id="1">
log: document.all.item("1") => <div id="1">
```

**Edge**
```
log: document.all["1"] => <head>
log: document.all("1") => <head>
log: document.all.item("1") => <div id="1">
```

**Chrome**
```
log: document.all["1"] => <head>
log: document.all("1") => <head>
log: document.all.item("1") => <head>
```

foolip replied:
Do you mean you'd be willing to drop support for `document.all("1")` from Edge? Presumably only after measuring the usage?

DigiTec commented:
> Hmm, I don't quite understand what you mean here. HTMLCollection and NodeList are pretty well-defined, mainly because they don't have any API surface that accepts a nameOrIndex

It does, but Edge HTMLAllCollection shares code with HTMLCollection, so we are going to align all of our behaviors on both of these objects. Including fixing our existing bad item behaviors on HTMLCollection (without the prose that exists for HTMLAllCollection)

DigiTec commented:
BTW - my last comment is above. If we decide and @bzbarsky agrees that we'll all do the Chrome behavior for legacycaller/item then I'm LGTM. If we align to FireFox instead then that LGTM. The reason for picking FireFox here is that the item("1") case is shared.

bzbarsky commented:
I'm confused by the sharing bit.  item() on HTMLCollection and HTMLAllCollection have totally different behavior: on HTMLCollection, item("abc") returns the first element of the list while on HTMLAllCollection it does something quite different, both in terms of how the value is treated and in terms of what the return type can be...

DigiTec commented:
i can test again later but we have overloads of htmlcollection internally. JS type is the same but internally divergent. So it may depend on which kind of htmlcollection you tested on.

foolip commented:
@DigiTec, @bzbarsky, can you review up to and including 9faf1ad? I don't really understand how we would get rid of the string-as-index handling in prose, and if we do have that, it seems more rational to have *only* that handling.

Unless I've overlooked something, which I probably have, it now holds that `document.all[x] == document.all(x) == document.item(x)` for any string or number *x*, but perhaps not for some funny objects with `toString` and `toValue`.

DigiTec commented:
If this simple case is right, then Chrome seems to already support the behavior you specified, unless there are some edge cases. I've only tested [-2...1] inclusive and a named string.
https://jsfiddle.net/j4o5zaLg/

This puts a nail in the coffin of HTMLAllCollection ever being the same as HTMLCollection. I think we'd like to believe that all is not special other than the truthy/falsy undefined behavior for `if (document.all)` but it clearly is. At this juncture I'd be okay with making this change. Chrome has precedent for the behavior, FireFox is only off in one case and Edge is off in two.

For @bzbarsky the test case also includes cases for HTMLCollection (getElementsByTagName) per our previous conversations. Take Edge with -2 in the getter, string getter, item as number and string as well.

I could live with this behavior.

bzbarsky commented:
> the test case also includes cases for HTMLCollection (getElementsByTagName)

Huh.  Yeah, that's totally not following what the spec has said for HTMLCollection for a long time now... :(

domenic commented:
It sounds like there's rough consensus on this patch now. We can in the future consider simplifications, if use counters reveal negligible amount of web code going down the more complex and undesirable paths (such as `undefined` or numeric strings).

I've squashed and rebased into a single commit. I plan to merge it tomorrow unless discussion re-opens between now and then.

bzbarsky commented:
If we achieve interop on the more complex thing, then there is zero incentive to change at that point, no matter what use counters say.

bzbarsky commented:
Is there a link to a version of the spec with the change merged in somewhere, by the way?  It's really hard (at least for me) to evaluate a change like this from a diff...

domenic commented:
@bzbarsky https://dl.dropboxusercontent.com/u/20140634/fix-htmlallcollection/index.html#the-htmlallcollection-interface should do the trick

bzbarsky commented:
Thanks, that's very helpful.  So we don't want to use (DOMString or unsigned long) for the legacycaller/item because that would not treat -2 as a string, right?  Might be worth explicitly pointing that out in a comment in the IDL.

Past that, I still don't think "all-indexed element" is good terminology, as I pointed out in https://github.com/whatwg/html/issues/775#issuecomment-192803891

Other than that, I can live with this, I guess.  We should make sure to add exhaustive web platform tests.

domenic commented:
It sounds like it might be best to just delegate to Web IDL's array index property name instead of redefining it and saying "this matches". Will push a commit for that.

bzbarsky commented:
The problem is that Web IDL's thing is an algorithm that takes an ES value as input, not an IDL string, which is what we have to work with here...

domenic commented:
_flips tables_ can you think of any linguistic tricks to allow us to reuse the algorithm? "_nameOrIndex_'s corresponding JavaScript value is an array index property name", perhaps? Or should we invent a new concept?

bzbarsky commented:
I guess we could invoke http://heycam.github.io/webidl/#DOMString-to-es and then send that through <http://heycam.github.io/webidl/#dfn-array-index-property-name>: none of that has side-effects and in practice UAs can do something simpler/saner.

It's totally dumb that this is an issue, I agree.

foolip commented:
I wouldn't really object to merging this, but also don't think it's that valuable if no implementor intends to proceed until use counter data (or similar) is available. Conflicts will have to be resolved, but it's also nice to not flip-flop the spec more than necessary.

To commit myself to some interpretation of the use counter data in advance, I'd say that code paths showing &ge;0.0001% usage on most (or all) days can stay, and those that round to zero on most (or all) days can go. This is entirely arbitrary, and based on the fact that chromestatus.com rounds usage to 0.0001% increments.

domenic commented:
OK, let's tag as "do not merge yet" until use counter data comes back? @DigiTec, would that be OK with you, or would you prefer to have something more-implementable in the spec for now?

foolip commented:
That was my thinking, but immediate implementor interest should trump everything else IMHO.

DigiTec commented:
I'd prefer we commit so that I can implement. At this point there are very few things we would plan on bringing back and those would be simple changes in edge cases for us to improve on in our second iteration.

To enumerate, the user counters are for:
1. optional behavior of the parameters in some cases, minor
2. subIndex behavior, could be major, but willing to flight it.

domenic commented:
Sounds good. I'll wait another couple hours for any further reviews or objections to surface, but my read is that everyone has more or less signed off at this point, so I'll merge soon.

foolip commented:
No objections, I'll simply revisit once use counter data is in.

bzbarsky commented:
What are we trying to usecounter, exactly?

smaug---- commented:
> ...immediate implementor interest should trump everything else IMHO.

(Just a random comment, not about this particular issue) That has led to problems so many times, and fixing later has proved to be hard. Though, perhaps we've got a bit better for not shipping unstable stuff.

foolip commented:
> What are we trying to usecounter, exactly?

See https://chromium.googlesource.com/chromium/src/+/69256151c9e821963929cf29a778b5df66027023/third_party/WebKit/LayoutTests/fast/dom/document-all-use-counters.html

The important one to proceed in Blink is `DocumentAllLegacyCallTwoArguments` which hasn't even been considered for the spec. The others are to check which of `document.all()`, `document.all(0)`, `document.all('0')`, `document.all.item()` and `document.all.item('0')` are really needed for compat.

foolip commented:
>> ...immediate implementor interest should trump everything else IMHO.

> (Just a random comment, not about this particular issue) That has led to problems so many times, and fixing later has proved to be hard. Though, perhaps we've got a bit better for not shipping unstable stuff.

I don't mean that implementor interest should trump everything on all issues, but in this instance it merely means *not* waiting for use counter data and instead going with (almost) the union of non-throwing behavior.
